lib/Arch/Arch.cpp

```diff
@@ -0,0 +1,185 @@
+      if (width <= 8) {
+        size_constraint = kMinBit8;
+      } else if (width <= 16) {
+        size_constraint = kMinBit16;
+      } else if (width <= 32) {
+        size_constraint = kMinBit32;
+      } else if (width <= 64) {
+        size_constraint = kMinBit64;
+      } else if (width <= 80) {
+        size_constraint = kMinBit80;
+      } else if (width <= 128) {
+        size_constraint = kMinBit128;
+      } else {
+        LOG(FATAL) << "Integer too big: "
+                   << remill::LLVMThingToString(&derived);
+      }
+      break;
+    }
+    case llvm::Type::FloatTyID: {
+      type_constraint = kTypeFloat;
+      // We automatically know it is 32-bit IEEE floating point type
+      size_constraint = kMinBit32;
+      break;
+    }
+    case llvm::Type::DoubleTyID: {
+      type_constraint = kTypeFloat;
+      // We automatically know it is 64-bit IEEE floating point type
+      size_constraint = kMinBit64;
+      break;
+    }
+    case llvm::Type::PointerTyID: {
+      type_constraint = kTypeIntegral;
+      size_constraint = kMinBit64;
+      break;
+    }
+    case llvm::Type::X86_FP80TyID: {
+      type_constraint = kTypeIntegral;
+      size_constraint = kMinBit80;
+      break;
+    }
+    case llvm::Type::VectorTyID: {
+      type_constraint = kTypeFloatOrVec;
+      size_constraint = kMinBit80;
+      break;
+    }
+    default: {
+      LOG(FATAL) << "Could not assign type and size constraints in "
+                    "TryRegisterAllocate()";
+      // TODO(aty): Handle other types like X86_MMXTyID, etc.
+      break;
+    }
+  }
+
+  for (size_t i = 0; i < register_constraints.size(); i++) {
+    if (reserved[i]) {
+      continue;
+    }
+
+    const RegisterConstraint &constraint = register_constraints[i];
+    // Iterate through the different sizes of a single register to find the
+    // smallest possible match
+    for (auto const &variant : constraint.variants) {
+      if (size_constraint & variant.size_constraint &&
+          type_constraint & variant.type_constraint) {
+        auto reg = arch->RegisterByName(variant.register_name);
+        if (!reg) {
+          LOG(FATAL) << "Could not find the register";
+        }
+        reserved[i] = true;
+        return reg;
+      }
+    }
+  }
+  return nullptr;
+}
+
+// For each element of the struct, try to allocate it to a register, if all of
+// them can be allocated, then return that allocation. Otherwise return a
+// nullptr.
+std::unique_ptr<std::vector<anvill::ValueDecl>>
+CallingConvention::TryReturnThroughRegisters(
+    const llvm::StructType &st,
+    const std::vector<RegisterConstraint> &constraints) {
+  auto ret = std::make_unique<std::vector<anvill::ValueDecl>>();
+  std::vector<bool> reserved(constraints.size(), false);
+  for (unsigned i = 0; i < st.getNumElements(); i++) {
+    anvill::ValueDecl value_decl = {};
+    auto reg =
+        TryRegisterAllocate(*st.getElementType(i), reserved, constraints);
```

Comment:
What if it is a struct in a struct?

Reply:
Or a struct in a vector, or what not. I think what might help is a kind of visitor, e.g. could be a function taking a type and a callback, and recursively descending down the type hierarchy until a "basic" type is hit.

Reply:
Handled the case of a struct split in e40ab35. However, I may need to do a minor bit of rearchitecting because arrays are split and vectors are split and packed.

Reply:
Is this resolved?

Reply:
Yes, fixed in 7db1b6f